### DISCUSSION TITLE
Improved "with keyword" with nested assignment / referential paths

### DISCUSSION THREAD

**Discussion by dellamonica** [dotnet/csharplang#8289] [Language Ideas]
Make the with keyword more powerful by allowing it to operate on nested properties.
For instance, let A be a record type containing another record Type B under property PropB.

```csharp
instanceOfA = instanceOfA with { PropB.AnotherProp = "changed" } 
```

This would be lowered to
```csharp
instanceOfA = instanceOfA with { PropB = instanceofA.PropB with { AnotherProp  = "changed" } } 
```

A related feature that could be useful is to include a reference to the instance being modified within the brackets.

```csharp
instanceOfA = instanceOfA with { Counter = this.Counter + 1 }
```

Though arguably even better (for such cases) would be 
```csharp
instanceOfA = instanceOfA with { Counter += 1 }
```

The *this* inside the brackets refer to the instanceOfA before assignment. Although in principle it could break some code where *this* is being used to refer to the instance of the containing class/struct, this type of code is probably very rare (indeed, there are very few cases where one needs to type out "this" when accessing local properties). Alternatively, the compiler could be smart enough to detect ambiguities and issue a warning, falling back to *this* referring to the containing type as legacy behavior.

**Comment by HaloFour** [comment]
See: #5176

**Reply by dellamonica** [reply]
Although that is a nice proposal, it is not a duplicate.

**Reply by HaloFour** [reply]
I didn't say that it was, although that proposal does cover nearly everything requested here.  The one thing it doesn't is the following syntax:

```cs
instanceOfA = instanceOfA with { Counter = this.Counter + 1 }
```

However, this is [already legal C# syntax](https://sharplab.io/#v2:D4AQzABATgpgxgeygEwgQQgbwLACgIETgQCWAdgC4QDCCArpTFFhAOYwUDcpZJXEAXzxDceYiABMNLHkJFI5KrQYUmLdvwDOHbiLmzCxDAFkAFBnKaKAQzJwYAeQBmaAJQz8cwwHYeV2/bOGADufAAWHl5RNPSMzAC8EBRhJJoAdMpxBlECnNmCwkA==) so I don't think it would be possible to interpret as suggested here.

**Reply by dellamonica** [reply]
Your short reply gave me the impression that you believed it to be a dupe.

Also, I don't see anything in the other proposal about nested records, which is the main proposal here. It would be nice if it could be combined with that proposal.

**Reply by HaloFour** [reply]
I think that was #77 and I think that was subsumed by #5176, but I could be wrong.

As mentioned in that linked discussion it would be a nice-to-have with property patterns also.

**Reply by dellamonica** [reply]
Yeah, #77 does cover the nested properties (I think I missed because it is in the "General" category and I was browsing "Language ideas"). Searching for "with" is not very helpful as you can imagine.
